src/runtime/virtcontainers/fs_share_linux.go

```diff
@@ -239,23 +254,81 @@ func (f *FilesystemShare) ShareFile(ctx context.Context, c *Container, m *Mount)
 	if !caps.IsFsSharingSupported() {
 		f.Logger().Debug("filesystem sharing is not supported, files will be copied")
 
-		fileInfo, err := os.Stat(m.Source)
-		if err != nil {
-			return nil, err
-		}
+		var ignored bool
+		srcRoot := filepath.Clean(m.Source)
 
-		// Ignore the mount if this is not a regular file (excludes
-		// directory, socket, device, ...) as it cannot be handled by
-		// a simple copy. But this should not be treated as an error,
-		// only as a limitation.
-		if !fileInfo.Mode().IsRegular() {
-			f.Logger().WithField("ignored-file", m.Source).Debug("Ignoring non-regular file as FS sharing not supported")
-			return nil, nil
+		walk := func(srcPath string, d fs.DirEntry, err error) error {
+
+			if err != nil {
+				return err
+			}
+
+			info, err := d.Info()
+			if err != nil {
+				return err
+			}
+
+			if !(info.Mode().IsRegular() || info.Mode().IsDir() || (info.Mode()&os.ModeSymlink) == os.ModeSymlink) {
+				f.Logger().WithField("ignored-file", srcPath).Debug("Ignoring non-regular file as FS sharing not supported")
+				if srcPath == srcRoot {
+					// Ignore the mount if this is not a regular file (excludes socket, device, ...) as it cannot be handled by
+					// a simple copy. But this should not be treated as an error, only as a limitation.
+					ignored = true
+					return filepath.SkipDir
+				}
+				return nil
+			}
+
+			dstPath := filepath.Join(guestPath, srcPath[len(srcRoot):])
+			f.Logger().Infof("ShareFile: Copying file from src (%s) to dest (%s)", srcPath, dstPath)
+			//TODO: Improve the agent protocol, to handle the case for existing symlink.
+			// Currently for an existing symlink, this will fail with EEXIST.
+			err = f.sandbox.agent.copyFile(ctx, srcPath, dstPath)
+			if err != nil {
+				f.Logger().WithError(err).Error("Failed to copy file")
+				return err
+			}
+
+			// Add fsNotify watcher for volume mounts
+			if strings.Contains(srcPath, "kubernetes.io~configmap") ||
```

Comment:
Should we have tighter checks here? That is, checks the paths documented in lines 302-305 below.

Because otherwise we could have a file that happens to contain say `kubernetes.io~configmap` in its name, and it will get unexpected treatment. User may want them to copied once at start but now they're being updated unintentionally.

Reply:
A possible tighter check was  discussed here.
https://github.com/kata-containers/kata-containers/pull/6342#discussion_r1151410817

Reply:
My read of this comment is that you had two points: optimising the check and making it stricter, but only the former was addressed.

Let's discuss the stricter check now then :)

Reply:
> Possibly we can do more strict checking here to reduce false-positives like this.
> https://go.dev/play/p/JNNXZSmzLkS

I meant this one cloud be a tighter check because it does not rely on `strings.Contains`.

Reply:
I talked with @bpradipt about this.

This PR is cherry-picking commits from the CCv0 branch. The original commits are tested as a part of CCv0 for some time, and I think it is OK to merge the proven commits as they are into the main for now. Then, we will submit another PR to improve the code with the stricter checking.

@wedsonaf what do you think?

Reply:
thanks @wedsonaf